### DISCUSSION TITLE
Authorization on a Template works in OpenAPI but 401 Unauthorized on static path

### DISCUSSION THREAD

**Discussion by agropper** [fastapi/fastapi#8891] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [ ] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [ ] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [ ] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

I've followed the tutorial with great success but stuck on how to force authorization on a pair of static pages. What's odd is that authorization works in the https://ognap-xf7vq.ondigitalocean.app/docs but returns 401 Unauthorized to https://ognap-xf7vq.ondigitalocean.app/items/foo (Sign in with ID "johndoe" and PW "secret").

**Do I need to create a real sign-in form at https://ognap-xf7vq.ondigitalocean.app/token or is there a simpler way to present the JWT via the browser?** There must be examples of this for FastAPI but I'm not finding them.

FYI - My goal is to demonstrate the difference between OAuth2 and IETF GNAP standards work in progress. See: https://www.ietf.org/archive/id/draft-ietf-gnap-core-protocol-05.html#name-compared-to-oauth-20 

The code is at https://github.com/agropper/OGNAP

**Comment by gyKa** [comment]
It depends on your needs. You can get your JWT token from /docs endpoint only once and then reuse it before any demonstration. Or you can create endpoint, which will accept your credentials and return JWT token.

I think the simplest way would be to use API client, such as [Postman](https://learning.postman.com/docs/sending-requests/authorization/#bearer-token) or [Insomnia](https://support.insomnia.rest/article/174-authentication) (or even [curl](https://nieldw.medium.com/using-curl-to-authenticate-with-jwt-bearer-tokens-55b7fac506bd)).

**Comment by agropper** [comment]
I think I understand that I need to trade a credential for a JWT somehow using something like a web browser as the user agent / client and FastAPI as the server (but using a templated form rather than the /docs endpoint. That's how I request the JWT. I could, for example have it sent to me with Twilio / SendGrid. 

This issue is because I'm unsure how to **present** the JWT to FastAPI that is protecting a static web page like a verifiable credential or QR code. It's a very newbie problem that I'm struggling with because it's hard for me as a novice to understand the relationship between FastAPI, Dependencies, and Templates.

**Comment by ricky-sb** [comment, marked as answer]
>This issue is because I'm unsure how to present the JWT to FastAPI  

Simple. Once you have your token, all your API calls need to have the `Authorization` header. How do you add the `Authorization` header? A few ways:
* Use an API client like [Postman](https://learning.postman.com/docs/sending-requests/authorization/#bearer-token)
* Use cURL in your Terminal. Example:
```bash
curl -X 'GET' \
  'https://ognap-xf7vq.ondigitalocean.app/items/foo' \
  -H 'accept: text/html' \
  -H 'Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJqb2huZG9lIiwiZXhwIjoxNjIzMjI0NDkyfQ.iDP5SDSbcBBq3GWHmgDBq1WqUp9brhbxLLTlUNb9R20'
```
* Use a frontend framework like React or Vue.js [to make your API calls](https://stackoverflow.com/questions/43051291/attach-authorization-header-for-all-axios-requests).

----
What you can't do:
* Directly visit this URL in Chrome, because Chrome won't automatically add that Authorization header or token for you

**Comment by agropper** [comment]
That’s useful but not the silly problem I’m having. As the title of the issue suggests, my problem is that the static page route is served by FastAPI even without the Authorization header. 

How do I make FastAPI require the Authorization header with the JWT?

**Comment by frankie567** [comment]
I think there is a misunderstanding there. FastAPI is first and foremost designed to build REST API; rather than server-side rendered websites like you would do with Django.

"Authorization" header authentication is designed for API; where you make sure for each request to manually set this header with your valid token. This is something very easy to do in `cURL`, Postman or even JavaScript. Nowadays, lot of websites are built with JavaScript frontend frameworks (like React or Vue.js) calling REST API programmatically.

However, when you deal with a rendered HTML page that is loaded by a browser, the browser has no way to set this "Authorization" header by itself; it's not meant for that. For this case, the most common way that has been used for years are **cookies**: browsers are designed to automatically send the cookies they have in memory **at each request**.

So, if you want to render a static HTML page requiring authentication, you'll have to do the following:

* Instead of returning a token in a JSON response, `login_for_access_token` endpoint should **set** it in a cookie (see https://fastapi.tiangolo.com/advanced/response-cookies/#response-cookies). On successful authentication, the browser will save this cookie in its memory.
* For endpoints requiring authentication, you'll be able to retrieve this cookie containing the token. You can use this security scheme to help you: https://github.com/tiangolo/fastapi/blob/master/fastapi/security/api_key.py#L54-L71
Basically, you should just change it on `get_current_user`.

And that should work. However, please note that cookie authentication comes with some security challenges, especially [CSRF](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html).

**Comment by agropper** [comment]
I understand @frankie567 explanation for why I would need a cookie. 

However, my problem is that I don't understand how to protect a static file. I sort of expected something like 

```
> @app.get("/static/", response_class=HTMLResponse)
> async def read_item(request: Request, id: str, current_user: User = Depends(get_current_active_user)):
>     return current_user
```
to return "Not Authorized" if the cookie or some other authorization header was not there but instead I get the file in static/

I seem to be missing something elementary.

**Comment by frankie567** [comment]
I think the problem comes from this:

```py
app.mount("/static", StaticFiles(directory="static"), name="static")
```

The point of this is to attach an "independent" router which only goal is to serve static files, without any particular logic behind.

If you define a FastAPI route under the same path (`/static`), it'll be catched by this router with no chance for your route to run.

**Comment by agropper** [comment]
If my route doesn’t run, how does the cookie or Authorization header get validated by the “independent” router?

**Comment by trottomv** [comment]
I have encountered a similar problem on a dependencies authorization of an endpoint, I've debug it and i notice that all the security classes that extrapolate "authorization" from the headers request have a problem. In `request.headers` we find the lowercase authorization key while the security classes execute `request.headers.get("Authorization")` with the titled key.

for example in `fastapi/security/http.py`:

```

class HTTPBase(SecurityBase):
    def __init__(
        self, *, scheme: str, scheme_name: Optional[str] = None, auto_error: bool = True
    ):
        self.model = HTTPBaseModel(scheme=scheme)
        self.scheme_name = scheme_name or self.__class__.__name__
        self.auto_error = auto_error

    async def __call__(
        self, request: Request
    ) -> Optional[HTTPAuthorizationCredentials]:
        authorization: str =   request.headers.get("Authorization")
        scheme, credentials = get_authorization_scheme_param(authorization)
        if not (authorization and scheme and credentials):
            if self.auto_error:
                raise HTTPException(
                    status_code=HTTP_403_FORBIDDEN, detail="Not authenticated"
                )
            else:
                return None
        return HTTPAuthorizationCredentials(scheme=scheme, credentials=credentials)
```

In my opinion should be:
`authorization: str = request.headers.get("authorization")` with lowercase "authorization" key.

**Comment by frankie567** [comment]
@trottomv I don't think it has to do with the issue. Besides, headers dictionary is definitely case-insensitive (see https://www.starlette.io/requests/#headers)

@agropper To be sure we understand well, do you want to:

* Render an HTML template with variables ?
* OR simply serve static files from the disk ?

**Comment by agropper** [comment]
Simply serve static files from the disk.

**Comment by agropper** [comment]
If you’re wondering why I’m going through this exercise, it’s because I am trying to model and understand the differences between OAuth2 and IETF GNAP as we work on that spec and in related issues in W3C. 

 https://www.ietf.org/archive/id/draft-ietf-gnap-core-protocol-05.html#name-compared-to-oauth-20

**Comment by frankie567** [comment]
Regarding static file serving ,`StaticFiles` should be used if you just wish to serve files without any custom logic. It's generally useful to serve CSS/JS or image files.

If you want to have authentication working, you'll need to implement an API route with the `get_current_user` dependency and manually return the file. `FileResponse` greatly helps in this matter (see https://fastapi.tiangolo.com/advanced/custom-response/#fileresponse).  Something like:

```py
@app.get("/static/{path:path}")
async def get_static_file(path: str, user = Depends(get_current_user)):
  return FileResponse(os.path.join("static", path))
```

The `:path` syntax is a convertor to ensure the parameter is handled correctly as a path. It's a feature coming from Starlette which can be useful in some cases. See https://fastapi.tiangolo.com/tutorial/path-params/#path-convertor and https://www.starlette.io/routing/#path-parameters

If the user is not authenticated, they will get a 401. Otherwise, the file will be served.

Full example: https://replit.com/@frankie567/fastapi-auth-static-files#app.py

**Comment by agropper** [comment]
@frankie567, I updated the route as you suggested but the file still gets served even tough I've not set a cookie or run the /token endpoint. https://ognap-xf7vq.ondigitalocean.app/static/QFVC.html 

Here's the code: https://github.com/agropper/OGNAP

**Comment by frankie567** [comment]
@agropper You need to remove that line:

```py
app.mount("/static", StaticFiles(directory="static"), name="static")
```

If you keep it, it'll take precedence over the `get_static_file` route.

**Comment by agropper** [comment]
Thanks. I removed and /static seems protected but the behavior is still a mystery.
https://ognap-xf7vq.ondigitalocean.app/static/QFVC.html and
https://ognap-xf7vq.ondigitalocean.app/static/QFVC-QR.png work differently in /docs
https://ognap-xf7vq.ondigitalocean.app/docs#/default/get_static_file_static__path__get

In some cases Authorization: Bearer is respected and in other cases it's not.

I'd like to avoid cookies and just gate on the Bearer JWT. Thanks for your help. As a novice, I just don't seem to be getting the picture for something that should be obvious.

The code is still at https://github.com/agropper/OGNAP

**Comment by frankie567** [comment]
I don't really understand what you mean by "Authorization: Bearer is respected and in other cases it's not.".

If a valid token is provided in `Authorization` header, the file will be served. Otherwise, you'll get a 401.

In the interactive documentation, if you click on the "Authorize" button at the top right, you'll get a window where you can input a username and a password. It'll automatically call your `/token` endpoint to get a fresh JWT and will automatically set the `Authorization` header for subsequent requests you'll make **within** the interactive documentation.

**Comment by agropper** [comment]
The behavior I experienced testing locally within the interactive
documentation was inconsistent with the file being returned even when the
Bearer header was not there. That could just be a cache problem, I guess,
but it’s worrisome.

What’s more important to my demonstration is to avoid cookies and use
Authorization: Bearer. I will be running different FastAPI instances in
different domains for the /static “Resource Server” vs. the token issuing
“Authorization Server”.

**Comment by agropper** [comment]
@frankie567 Thank you for your help! I've made a lot of progress toward understanding [the difference between OAuth2 and GNAP](https://www.ietf.org/archive/id/draft-ietf-gnap-core-protocol-05.html#name-compared-to-oauth-20). I'm now able to run two separate FastAPI servers on separate domains for the Authorization Server (AS) and the Resource Server (RS) and protect the (static) resources with a Bearer JWT issued by the Authorization Server.

To try it:
AS: https://ognap-xf7vq.ondigitalocean.app/ (Authenticate as "johndoe" with the password "secret") 
RS: https://ognap-rs-me9ms.ondigitalocean.app/static/QFVC.html (use curl or Postman to add the Bearer JWT to the Authenticate header)

Because the AS and RS are in separate domains by design, I now need to come up with a client app that can do this and start reading the GNAP specs for real. 

If anyone wants to take a look at my RS (and AS code) and suggest how I can make it simpler or clearer, that will be appreciated. Either way, I can close this issue in a few days.

Thank you!
